File: scripts/validate_seds.py
#!/usr/bin/env python3
"""
Project      : SMRITI Business OS
Product      : SMRITI Enterprise Design System (SEDS)
Script       : SEDS CI/CD Legacy UI Enforcement Linter
Author       : Jawahar Ramkripal Mallah
Version      : 1.0.0
Classification: Internal CI/CD Governance Linter
"""

import os
import re
import sys

TARGET_DIR = os.path.join(os.path.dirname(__file__), "..", "src", "components")

# Prohibited legacy Tailwind slate class patterns
PROHIBITED_PATTERNS = [
    re.compile(r'bg-slate-[0-9]+(/[0-9]+)?'),
    re.compile(r'text-slate-[0-9]+(/[0-9]+)?'),
    re.compile(r'border-slate-[0-9]+(/[0-9]+)?'),
    re.compile(r'ring-slate-[0-9]+(/[0-9]+)?'),
    re.compile(r'divide-slate-[0-9]+(/[0-9]+)?'),
    re.compile(r'from-slate-[0-9]+(/[0-9]+)?'),
    re.compile(r'to-slate-[0-9]+(/[0-9]+)?'),
    re.compile(r'via-slate-[0-9]+(/[0-9]+)?'),
]

# Excluded paths (e.g. documentation markdown or historical mocks if explicitly exempted)
EXCLUDED_SUBPATHS = [
    os.path.join("website"), # Public website product is an independent tier per AOP-002
]

def scan_file(filepath):
    violations = []
    rel_path = os.path.relpath(filepath, TARGET_DIR)
    
    # Check exclusions
    for exc in EXCLUDED_SUBPATHS:
        if exc in rel_path:
            return violations

    with open(filepath, "r", encoding="utf-8", errors="ignore") as f:
        for line_num, line in enumerate(f, 1):
            for pattern in PROHIBITED_PATTERNS:
                matches = pattern.findall(line)
                if matches or pattern.search(line):
                    match_str = pattern.search(line).group(0)
                    violations.append((rel_path, line_num, match_str, line.strip()))
    return violations

def main():
    print("======================================================================")
    print(" SMRITI Enterprise Design System (SEDS) CI/CD Governance Linter")
    print("======================================================================")
    
    total_violations = 0
    violating_files = set()
    
    abs_target = os.path.abspath(TARGET_DIR)
    print(f"Scanning target directory: {abs_target}\n")

    for root, _, files in os.walk(abs_target):
        for file in files:
            if file.endswith(".tsx") or file.endswith(".ts"):
                filepath = os.path.join(root, file)
                file_violations = scan_file(filepath)
                if file_violations:
                    violating_files.add(os.path.relpath(filepath, abs_target))
                    for rel_path, line_num, match_str, content in file_violations:
                        total_violations += 1
                        print(f"❌ [SEDS-VIOLATION] {rel_path}:{line_num} -> Found '{match_str}'")
                        print(f"   Snippet: {content[:100]}")

    print("\n----------------------------------------------------------------------")
    print(f"Total Prohibited Legacy Slate Violations: {total_violations}")
    print(f"Total Violating Files: {len(violating_files)}")
    print("----------------------------------------------------------------------")

    if total_violations > 0:
        print("\n❌ CI/CD RELEASE GATE FAILED: SEDS legacy token violations detected!")
        sys.exit(1)
    else:
        print("\n✅ CI/CD RELEASE GATE PASSED: Zero legacy UI slate violations found!")
        sys.exit(0)

if __name__ == "__main__":
    main()

File: scripts/validate_seds.js
/**
 * Project      : SMRITI Business OS
 * Product      : SMRITI Enterprise Design System (SEDS)
 * Script       : SEDS CI/CD Legacy UI Enforcement Linter (Node.js)
 * Author       : Jawahar Ramkripal Mallah
 * Version      : 1.0.0
 * Classification: Internal CI/CD Governance Linter
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const TARGET_DIR = path.join(__dirname, '..', 'src', 'components');

const PROHIBITED_PATTERNS = [
  /bg-slate-[0-9]+(\/[0-9]+)?/g,
  /text-slate-[0-9]+(\/[0-9]+)?/g,
  /border-slate-[0-9]+(\/[0-9]+)?/g,
  /ring-slate-[0-9]+(\/[0-9]+)?/g,
  /divide-slate-[0-9]+(\/[0-9]+)?/g,
  /from-slate-[0-9]+(\/[0-9]+)?/g,
  /to-slate-[0-9]+(\/[0-9]+)?/g,
  /via-slate-[0-9]+(\/[0-9]+)?/g,
];

const EXCLUDED_SUBPATHS = [
  'website', // Public marketing website product is an independent product tier per AOP-002
];

function scanDirectory(dir, fileList = []) {
  const files = fs.readdirSync(dir);
  for (const file of files) {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);
    if (stat.isDirectory()) {
      scanDirectory(filePath, fileList);
    } else if (file.endsWith('.tsx') || file.endsWith('.ts')) {
      fileList.push(filePath);
    }
  }
  return fileList;
}

function main() {
  console.log('======================================================================');
  console.log(' SMRITI Enterprise Design System (SEDS) CI/CD Governance Linter');
  console.log('======================================================================');

  const files = scanDirectory(TARGET_DIR);
  let totalViolations = 0;
  const violatingFiles = new Set();

  for (const filePath of files) {
    const relPath = path.relative(TARGET_DIR, filePath);
    if (EXCLUDED_SUBPATHS.some((exc) => relPath.startsWith(exc))) {
      continue;
    }

    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.split('\n');

    lines.forEach((line, index) => {
      PROHIBITED_PATTERNS.forEach((pattern) => {
        pattern.lastIndex = 0;
        const matches = line.match(pattern);
        if (matches) {
          matches.forEach((match) => {
            totalViolations++;
            violatingFiles.add(relPath);
            console.log(`❌ [SEDS-VIOLATION] ${relPath}:${index + 1} -> Found '${match}'`);
            console.log(`   Snippet: ${line.trim().substring(0, 100)}`);
          });
        }
      });
    });
  }

  console.log('\n----------------------------------------------------------------------');
  console.log(`Total Prohibited Legacy Slate Violations: ${totalViolations}`);
  console.log(`Total Violating Files: ${violatingFiles.size}`);
  console.log('----------------------------------------------------------------------');

  if (totalViolations > 0) {
    console.log('\n❌ CI/CD RELEASE GATE FAILED: SEDS legacy token violations detected!');
    process.exit(1);
  } else {
    console.log('\n✅ CI/CD RELEASE GATE PASSED: Zero legacy UI slate violations found!');
    process.exit(0);
  }
}

main();
